.circleci/config.yml

```diff
@@ -0,0 +1,154 @@
+    FIRRTL_BRANCH: master
+    FIRRTL_REV: master
+    SBT_ARGS: ""
+
+
+version: 2
+
+jobs:
+  build-prep:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+
+    # Perform a "default" checkout to get ssh set up.
+      - checkout:
+          path: chisel3
+
+      - run:
+          command: |
+            date > date.prep
+            printenv
+            ls -l
+            git clone --depth 10 --branch $FIRRTL_BRANCH "$FIRRTL_REPO" firrtl && (cd firrtl && git checkout $FIRRTL_REV)
+            echo $FIRRTL_REV && (cd firrtl && git log -1 > ../firrtl.log)
+
+
+      - persist_to_workspace:
+          root: /home/chisel
+          paths:
+            - repo
+            - .ivy2
+            - .m2
+            - .sbt
+
+  build-firrtl:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+      - attach_workspace:
+          at: /home/chisel
+
+      # publish FIRRTL
+      - run:
+          command: |
+            date > date.firrtl
+            (cd firrtl && cat /dev/null | sbt $SBT_ARGS +publishLocal)
+
+      - persist_to_workspace:
+          root: /home/chisel
+          paths:
+            - repo
+            - .ivy2
+            - .m2
+            - .sbt
+
+  # Define a pure build chisel - currently unused since we can compile and test chisel in one step
+  #  and we don't have downstream jobs that need the published version of chisel.
+  build-chisel:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+      - attach_workspace:
+          at: /home/chisel
+
+      # publish chisel
+      - run: date > date.chisel3 && (cd chisel3 && sbt $SBT_ARGS +publishLocal)
+
+      - persist_to_workspace:
+          root: /home/chisel
+          paths:
+            - repo
+            - .ivy2
+            - .m2
+            - .sbt
+
+  test-chisel-2.11:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+      - attach_workspace:
+          at: /home/chisel
+
+      # Set environment
+      - run: echo 'export PATH="/opt/verilator/verilator_3_922/bin:/opt/yosys/bin:$PATH"' >> $BASH_ENV
+
+      # We seem to have memory leaks in chisel tests -
+      # we run out of memory trying to "sbt +test" but a single "sbt test" passes.
+      - run:
+          command: |
+            (cd chisel3 && sbt $SBT_ARGS "++2.11.12" test)
+
+  test-chisel-2.12:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+      - attach_workspace:
+          at: /home/chisel
+
+      # Set environment
+      - run: echo 'export PATH="/opt/verilator/verilator_3_922/bin:/opt/yosys/bin:$PATH"' >> $BASH_ENV
+
+      # We seem to have memory leaks in chisel tests -
+      # we run out of memory trying to "sbt +test" but a single "sbt test" passes.
+      - run:
+          command: |
+            (cd chisel3 && sbt $SBT_ARGS "++2.12.6" test)
+
+  checkstyle-chisel:
+    <<: *defaultImageUserWDEnv
+
+    steps:
+      - attach_workspace:
+          at: /home/chisel
+
+      - run:
+          command: |
+            (cd chisel3 && sbt $SBT_ARGS scalastyle  test:scalastyle)
```

Comment:
Do you want to align this rule with the Jenkins "must have less than 40 warnings" in `sbt scalastyle`?

Reply:
Yes. Good point.